lib/storage/metadata/mongoclient/MongoClientInterface.js

```diff
@@ -1295,25 +1341,88 @@ class MongoClientInterface {
+                // Should not return an error when no object is found
+                if (err.is.NoSuchKey) {
+                    return cb(null);
+                }
                 log.error(
                     'deleteObjectNoVer: error deleting object with no version',
-                    { error: err.message });
+                    { error: err.message, bucket: bucketName, key: objName });
                 return cb(errors.InternalError);
             }
-            if (result.ok !== 1) {
+            if (res.ok !== 1) {
                 log.error(
-                    'deleteObjectNoVer: failed deleting object with no version',
-                    { error: err.message });
+                    'deleteObjectNoVer: failed to delete object',
+                    { bucket: bucketName, object: objName });
                 return cb(errors.InternalError);
             }
             return cb(null);
         });
     }
 
+    /**
+     * Flags the object before deleting it, this is done
+     * to keep object metadata in the oplog, as oplog delete
+     * events don't contain any object metadata
+     * @param {Object} collection MongoDB collection
+     * @param {string} bucketName bucket name
+     * @param {string} key Key of the object to delete
+     * @param {object} filter additional query filters
+     * @param {Logger}log logger instance
+     * @param {Function} cb callback containing error
+     * and BulkWriteResult
+     * @return {undefined}
+     */
+    internalDeleteObject(collection, bucketName, key, filter, log, cb) {
+        // filter used when finding and updating object
+        const findUpdateFilter = Object.assign({
+            _id: key,
+        }, filter);
+        // filter used when deleting object
+        const deleteFilter = Object.assign({
+            '_id': key,
+            'value.deleted': true,
+        }, filter);
+        async.waterfall([
+            next => collection.findOne(findUpdateFilter, {}, (err, doc) => {
+                if (err) {
+                    log.error('internalDeleteObject: error getting object',
+                        { bucket: bucketName, object: key, error: err.message });
+                    return next(errors.InternalError);
+                }
+                if (!doc) {
+                    log.error('internalDeleteObject: unable to find target object to delete',
+                        { bucket: bucketName, object: key });
+                    return next(errors.NoSuchKey);
+                }
+                const objMetadata = new ObjectMD(doc.value);
+                // adding delete flag to object metadata
+                objMetadata.setDeleted(true);
+                // updating originOp
+                objMetadata.setOriginOp('s3:ObjectRemoved:Delete');
+                return next(null, objMetadata.getValue());
+            }),
+            (objMetadata, next) => collection.bulkWrite([
+                {
+                    updateOne: {
```

Comment:
there may be some race condition here: e.g. if the document is updated after it has been read (line 1389) but before we re-set its metadata... and in this case we would reset some older (outdated) metadata...

this should not happen with version documents, which are immutable, but may be the case when updating the master (in non-versioned bucket).

→ i think we should add an extra condition, to ensure we only update/delete the right document.

Reply:
IMO Adding that extra condition will lead to adding more logic for handling what to do in case we don't find that version of the object (retry logic). I think that'll hinder the performance of deletion operations for little benefit.

I think deletion events should not be used to get volatile information such as tags and such, we always get update events for those types of operations, and those events should be the ones used instead.

Reply:
Also i think having deletion operations that occur at the same time as update ones should not be a common occurence so we should be good for the most part

Reply:
> I think deletion events should not be used to get volatile information such as tags and such, we always get update events for those types of operations, and those events should be the ones used instead.

Yes, i agree. But if the information is available, it will eventually be used (see [Hyrum's law](https://www.hyrumslaw.com/)) : so i think either we need to go the extra mile and protect against this, or we need to remove the mutable metadata...